File: app/src/main/java/de/hdc/kspchecklist/ItemListActivity.java
package de.hdc.kspchecklist;

import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.text.InputType;
import android.view.ContextMenu;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.EditText;
import android.widget.ListView;

import java.util.ArrayList;
import java.util.List;

import de.hdc.kspchecklist.data.DataIO;
import de.hdc.kspchecklist.data.ListItem;

/**
 * An activity representing a list of Items. This activity
 * has different presentations for handset and tablet-size devices. On
 * handsets, the activity presents a list of items, which when touched,
 * lead to a {@link ItemDetailActivity} representing
 * item details. On tablets, the activity presents the list of items and
 * item details side-by-side using two vertical panes.
 */
public class ItemListActivity extends AppCompatActivity implements AdapterView.OnItemClickListener {

    public final static String DETAIL_MESSAGE = "DETAIL_MESSAGE";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        SharedPreferences sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this);
        Boolean dark = sharedPrefs.getBoolean(getString(R.string.preference_theme_key), true);
        if (dark) {
            setTheme(R.style.AppTheme);
        } else {
            setTheme(R.style.AppThemeLight);
        }
        setContentView(R.layout.activity_item_list);
        Toolbar myToolbar = (Toolbar) findViewById(R.id.list_toolbar);
        setSupportActionBar(myToolbar);

        /**
         * If installed and run for the first time, copy assets files to local storage.
         */
        SharedPreferences wmbPreference = PreferenceManager.getDefaultSharedPreferences(this);
        boolean isFirstRun = wmbPreference.getBoolean("FIRS_TRUN", true);
        if (isFirstRun) {
            // Code to run once
            SharedPreferences.Editor editor = wmbPreference.edit();
            editor.putBoolean("FIRSTRUN", false);
            editor.apply();

            DataIO.copyAssetsFiles(getApplicationContext());
        }

//        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
//        fab.setOnClickListener(new View.OnClickListener() {
//            @Override
//            public void onClick(View view) {
//                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
//                        .setAction("Action", null).show();
//            }
//        });

        // Construct the data source
        items = new ArrayList<>();
        List<String> list = DataIO.getDirList(this.getApplicationContext());
        for (String s : list) {
            items.add(ListItem.create(s));
        }
        // Create the adapter to convert the array to views
        adapter = new ItemListAdapter(this, items);

        ListView listView = (ListView) findViewById(R.id.item_list);
        assert listView != null;
        listView.setAdapter(adapter);
        listView.setChoiceMode(ListView.CHOICE_MODE_SINGLE);
        listView.setOnItemClickListener(this);
        registerForContextMenu(listView);
    }

    @Override
    public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
          Intent intent = new Intent(this, ItemDetailActivity.class);
        intent.putExtra(DETAIL_MESSAGE, items.get(i).name);
        startActivity(intent);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu items for use in the action bar
        MenuInflater inflater = getMenuInflater();
        inflater.inflate(R.menu.list_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_settings:
                startActivity(new Intent(this, SettingsActivity.class));
                return true;

            case R.id.action_list_add:
                AlertDialog.Builder builder = new AlertDialog.Builder(this);
                builder.setTitle("New checklist");

                final EditText input = new EditText(this);
                input.setInputType(InputType.TYPE_CLASS_TEXT);
                builder.setView(input);
                builder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        String name = input.getText().toString();
                        DataIO.createLocalFile(getApplicationContext(), name + ".txt");
                        items.add(ListItem.create(name));
                        adapter.notifyDataSetChanged();
                        Intent intent = new Intent(getApplication(), ItemDetailActivity.class);
                        intent.putExtra(DETAIL_MESSAGE, name);
                        startActivity(intent);
                    }
                });
                builder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.cancel();
                    }
                });

                builder.show();

                return true;

            default:
                // If we got here, the user's action was not recognized.
                // Invoke the superclass to handle it.
                return super.onOptionsItemSelected(item);

        }
    }

    @Override
    public void onCreateContextMenu(final ContextMenu menu, final View v, final ContextMenu.ContextMenuInfo menuInfo) {
        if (v.getId()==R.id.item_list) {
            AdapterView.AdapterContextMenuInfo info = (AdapterView.AdapterContextMenuInfo) menuInfo;
            menu.setHeaderTitle("Checklist: " + items.get(info.position));
            String[] menuItems = getResources().getStringArray(R.array.menu);
            for (int i = 0; i<menuItems.length; i++) {
                menu.add(Menu.NONE, i, i, menuItems[i]);
            }
        }
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {
        final AdapterView.AdapterContextMenuInfo info = (AdapterView.AdapterContextMenuInfo)item.getMenuInfo();
        switch (item.getItemId()) {
            case 0: {  // edit
                AlertDialog.Builder builder = new AlertDialog.Builder(this);
                builder.setTitle("New name");

                final EditText input = new EditText(this);
                input.setInputType(InputType.TYPE_CLASS_TEXT);
                builder.setView(input);
                builder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        String name = input.getText().toString();
                        DataIO.renameLocalFile(getApplicationContext(), items.get(info.position).name + ".txt", name + ".txt");
                        items.set(info.position, ListItem.create(name));
                        adapter.notifyDataSetChanged();
                    }
                });
                builder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.cancel();
                    }
                });

                builder.show();
                return true;
            }
            case 1: {  // delete
                AlertDialog.Builder builder = new AlertDialog.Builder(this);
                builder.setTitle("Delete " + items.get(info.position))
                        .setMessage("Do you really want to delete this checklist?");
                builder.setPositiveButton("OK", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        DataIO.deleteLocalFile(getApplicationContext(), items.get(info.position).name + ".txt");
                        items.remove(info.position);
                        adapter.notifyDataSetChanged();
                    }
                });
                builder.setNegativeButton("Cancel", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.cancel();
                    }
                });

                builder.show();
                return true;
            }
            default: return false;
        }
    }

    private ArrayList<ListItem> items;
    private ItemListAdapter adapter;

}

File: app/src/main/java/de/hdc/kspchecklist/ItemListActivity.kt
package de.hdc.kspchecklist

import android.content.*
import android.os.*
import android.preference.*
import android.text.*
import android.view.*
import android.widget.*
import androidx.appcompat.app.*
import com.google.firebase.analytics.*
import de.hdc.framework.*
import de.hdc.kspchecklist.data.*
import de.hdc.kspchecklist.domain.*
import de.hdc.kspchecklist.framework.*
import kotlinx.android.synthetic.main.activity_item_list.*
import java.util.*

private const val FIRST_RUN = "FIRSTRUN"

/**
 * An activity representing a list of Items. This activity
 * has different presentations for handset and tablet-size devices. On
 * handsets, the activity presents a list of items, which when touched,
 * lead to a [ItemDetailActivity] representing
 * item details. On tablets, the activity presents the list of items and
 * item details side-by-side using two vertical panes.
 */
class ItemListActivity : AppCompatActivity(), AdapterView.OnItemClickListener {

    private lateinit var items: ArrayList<CheckList>
    private lateinit var adapter: ItemListAdapter
    private val persistence: CheckListPersistenceSource by lazy {
      CheckListPersistenceImpl(applicationContext)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val sharedPrefs = PreferenceManager.getDefaultSharedPreferences(this)
        val dark = sharedPrefs.getBoolean(getString(R.string.preference_theme_key), true)
        if (dark) {
            setTheme(R.style.AppTheme_NoActionBar)
        } else {
            setTheme(R.style.AppThemeLight)
        }
        setContentView(R.layout.activity_item_list)
        setSupportActionBar(list_toolbar)

        // Obtain the FirebaseAnalytics instance.
        val mFirebaseAnalytics = FirebaseAnalytics.getInstance(this)
        if (BuildConfig.DEBUG) {
            val params = Bundle()
            params.putString("app_started", "true")
            mFirebaseAnalytics.logEvent("share_image", params)
        }

        /**
         * If installed and run for the first time, copy assets files to local storage.
         */
        val wmbPreference = PreferenceManager.getDefaultSharedPreferences(this)
        val isFirstRun = wmbPreference.getBoolean(FIRST_RUN, true)
        if (isFirstRun) {
            // Code to run once
            val editor = wmbPreference.edit()
            editor.putBoolean(FIRST_RUN, false)
            editor.apply()

            persistence.copyAssetsFiles()
        }

        //        FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);
        //        fab.setOnClickListener(new View.OnClickListener() {
        //            @Override
        //            public void onClick(View view) {
        //                Snackbar.make(view, "Replace with your own action", Snackbar.LENGTH_LONG)
        //                        .setAction("Action", null).show();
        //            }
        //        });

        // Construct the data source
        items = persistence.getLists()
        items.sort()
        // Create the adapter to convert the array to views
        adapter = ItemListAdapter(this, items)

        item_list.adapter = adapter
        item_list.choiceMode = ListView.CHOICE_MODE_SINGLE
        item_list.onItemClickListener = this
        registerForContextMenu(item_list)
    }

    override fun onItemClick(adapterView: AdapterView<*>, view: View, i: Int, l: Long) {
        val intent = Intent(this, ItemDetailActivity::class.java)
        intent.putExtra(DETAIL_MESSAGE, items[i].name)
        startActivity(intent)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu items for use in the action bar
        val inflater = menuInflater
        inflater.inflate(R.menu.list_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.action_settings -> {
                startActivity(Intent(this, SettingsActivity::class.java))
                return true
            }

            R.id.action_list_add -> {
                val builder = AlertDialog.Builder(this)
                builder.setTitle(getString(R.string.new_checklist))

                val input = EditText(this)
                input.inputType = InputType.TYPE_CLASS_TEXT
                builder.setView(input)
                builder.setPositiveButton(getString(R.string.ok)) { _, _ ->
                    val name = input.text.toString()
                    persistence.addCheckList("$name.txt")
//                    DataIO.createLocalFile(applicationContext, "$name.txt")
                    items.add(CheckList(name))
                    items.sort()
                    adapter.notifyDataSetChanged()
                    val intent = Intent(application, ItemDetailActivity::class.java)
                    intent.putExtra(DETAIL_MESSAGE, name)
                    startActivity(intent)
                }
                builder.setNegativeButton(getString(R.string.cancel)) { dialog, _ -> dialog.cancel() }

                builder.show()

                return true
            }

            else ->
                // If we got here, the user's action was not recognized.
                // Invoke the superclass to handle it.
                return super.onOptionsItemSelected(item)
        }
    }

    override fun onCreateContextMenu(menu: ContextMenu, v: View, menuInfo: ContextMenu.ContextMenuInfo) {
        if (v.id == R.id.item_list) {
            val info = menuInfo as AdapterView.AdapterContextMenuInfo
            menu.setHeaderTitle("Checklist: " + items[info.position])
            val menuItems = resources.getStringArray(R.array.menu)
            for (i in menuItems.indices) {
                menu.add(Menu.NONE, i, i, menuItems[i])
            }
        }
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        val info = item.menuInfo as AdapterView.AdapterContextMenuInfo
        when (item.itemId) {
            0 -> {  // edit
                val builder = AlertDialog.Builder(this)
                builder.setTitle(getString(R.string.new_name))

                val input = EditText(this)
                input.inputType = InputType.TYPE_CLASS_TEXT
                builder.setView(input)
                builder.setPositiveButton(getString(R.string.ok)) { _, _ ->
                    val name = input.text.toString()
                    persistence.renameList(items[info.position].name + ".txt", "$name.txt")
                    items[info.position] = CheckList(name)
                    adapter.notifyDataSetChanged()
                }
                builder.setNegativeButton(getString(R.string.cancel)) { dialog, _ -> dialog.cancel() }

                builder.show()
                return true
            }
            1 -> {  // delete
                val builder = AlertDialog.Builder(this)
                builder.setTitle(items[info.position].toString())
                        .setMessage(getString(R.string.confirm_delete))
                builder.setPositiveButton(getString(R.string.ok)) { _, _ ->
                    persistence.deleteList(items[info.position].name + ".txt")
                    items.removeAt(info.position)
                    adapter.notifyDataSetChanged()
                }
                builder.setNegativeButton(getString(R.string.cancel)) { dialog, _ -> dialog.cancel() }

                builder.show()
                return true
            }
            else -> return false
        }
    }

    companion object {

        const val DETAIL_MESSAGE = "DETAIL_MESSAGE"
    }

}
